dex/src/main/java/io/crate/data/BatchIterator.java

```diff
@@ -58,22 +58,14 @@
  *
  * Concurrent usage of a BatchIterator is not supported.
  */
-public interface BatchIterator extends Killable {
+public interface BatchIterator<T> extends Killable {
 
     /**
-     * This method returns a columns object which can be used to access the underlying data of the current iterator
-     * position.
+     * This method returns the item that is at the current position of the BatchIterator.
      *
-     * Note that it is only valid to call {@link Input#value()} on any of the returned columns if the iterator is
-     * positioned on a valid row, which is only the case if the last call to {@link #moveNext()} returned true.
-     *
-     * This method is valid to be called over the whole lifetime of the iterator, regardless of the state of the
-     * iterator or its position. However it is good practice for consumers of this iterator to gather the columns
-     * before iterating. This method is required to always return the same object on every call.
-     *
-     * @return a columns object
+     * The behaviour of this method if the BatchIterator is "not on a position" is undefined.
```

Comment:
Why is it undefined and not throwing an exception for example?

Reply:
I'd say because it would require a check on every call to `currentElement()` but perhaps @mfussenegger can give more insight.

Reply:
We're not enforcing a behaviour in the contract because we're treating that as an iterator usage programming error (that would be exposed in tests)

Reply:
That it previously raised an error was part of the `Columns` implementations and it was kinda cheap (we just resetted the position, which we had to do anyway).

But now since the elements can be anything now this cheap failure behaviour is
gone. It would now require an additional state check.

I'll add a `SENTINEL` parameter to the `InMemoryBatchIterator`: That can be
used to provide a instance that should be returned by `currentElement()` in
case the BatchIterator is off-position. This should in practive have the same
effect that we had so far.